### ISSUE TITLE
Authen::U2F::Tester signature not recognize by Yubico libraries

### ISSUE THREAD

**Issue by guimard** [mschout/perl-authen-u2f-tester#2]
Hello,
I'm using Authen::U2F::Tester to test Crypt::U2F::Server. This library works with Yubikeys but when trying to use it with Authen::U2F::Tester, I've the following error (Yubico library in debug mode):
```
signatureData: AAAAAAEwRQIhAJGv06fg4BGb3dJvNKwcZuFWjsY3fq5yAAOjZGs0lSPzAiBrmkwyqMcGuXfmPyy4lL5Rh6zsrSxyYY6Hex61gP6udA
clientData: eyJjaGFsbGVuZ2UiOiJyNURuZ2FtaFBocGN6NW13RWN6emY4akd1ZHc1UFJ1akNBc1FCQ2p2UE9JIiwiY2lkX3B1YmtleSI6InVudXNlZCIsInR5cCI6Im5hdmlnYXRvci5pZC5nZXRBc3NlcnRpb24iLCJvcmlnaW4iOiJodHRwOi8vYXV0aC5leGFtcGxlLmNvbSJ9
keyHandle: MFAGCWCGSAFlAwQCAQQhAwWOjET9DigCAs_wLzGfMHiegPgNzxt5aous6cjaoMa2BCCLcu99O5fNZPfHuefqfotZYo6VhYEiHmK_Gw5JNzVv6w
signatureData Hex: 
00 00 00 00 01 30 45 02 21 00 91 af d3 a7 e0 e0 
11 9b dd d2 6f 34 ac 1c 66 e1 56 8e c6 37 7e ae 
72 00 03 a3 64 6b 34 95 23 f3 02 20 6b 9a 4c 32 
a8 c7 06 b9 77 e6 3f 2c b8 94 be 51 87 ac ec ad 
2c 72 61 8e 87 7b 1e b5 80 fe ae 74 
User presence byte mismatch
```
To reproduce it, download Crypt::U2F::Server 0.42, and apply this to enable falling test (and debug):
```
diff -r e8ff8e1704a9 lib/Crypt/U2F/Server/Simple.pm
--- a/lib/Crypt/U2F/Server/Simple.pm    Fri Mar 30 07:02:21 2018 +0200
+++ b/lib/Crypt/U2F/Server/Simple.pm    Mon Apr 09 12:11:55 2018 +0200
@@ -37,7 +37,7 @@
     my $self = bless \%config, $class;
 
     if(!$refCount) {
-        my $rc = Crypt::U2F::Server::u2fclib_init(0);
+        my $rc = Crypt::U2F::Server::u2fclib_init(1);
         if(!$rc) {
             $errstr = Crypt::U2F::Server::u2fclib_getError();
             return;
diff -r e8ff8e1704a9 t/12-full-test.t
--- a/t/12-full-test.t  Fri Mar 30 07:02:21 2018 +0200
+++ b/t/12-full-test.t  Mon Apr 09 12:11:55 2018 +0200
@@ -5,7 +5,7 @@
 
 use strict;
 use warnings;
-use constant TESTS => 11;
+use constant TESTS => 12;
 
 use Test::More tests => TESTS;
 
@@ -106,7 +106,7 @@
         }
     );
 
-    #ok( $crypter->authenticationVerify($sign), 'Good signature' )
-    #  or diag( Crypt::U2F::Server::Simple::lastError() );
+    ok( $crypter->authenticationVerify($sign), 'Good signature' )
+      or diag( Crypt::U2F::Server::Simple::lastError() );
     undef $crypter;
 }
```
Then launch:
`perl Makefile.PL && make && perl -Iblib/arch -Iblib/lib t/12-full-test.t`

The bug is revealed here (Yubico server library: `u2f-server/core.c`):
```
static u2fs_rc
parse_signatureData2(const unsigned char *data, size_t len,
                     uint8_t * user_presence, uint32_t * counter,
                     u2fs_ECDSA_t ** signature)
{
  /*
     +-----------------------------------+
     | 1 |     4     |      implied      |
     +-----------------------------------+
     user presence
     counter
     signature
   */

  int offset = 0;
  u2fs_rc rc;

  if (len <= 1 + U2FS_COUNTER_LEN) {
    if (debug)
      fprintf(stderr, "Length mismatch\n");
    return U2FS_FORMAT_ERROR;
  }

  *user_presence = data[offset++] & 0x01;

  if (*user_presence == 0) {
    if (debug)
      fprintf(stderr, "User presence byte mismatch\n");
    return U2FS_FORMAT_ERROR;
  }

  memcpy((char *) counter, data + offset, U2FS_COUNTER_LEN);

  offset += U2FS_COUNTER_LEN;

  size_t signature_len = len - offset;
  rc = decode_ECDSA(data + offset, signature_len, signature);

  if (rc != U2FS_OK) {
    return rc;
  }

  return U2FS_OK;
}
```

**Comment by guimard** [comment]
It means that Yubico libraries require that first byte of signature must have bit 0 set to 1

**Comment by mschout** [comment]
I don't have a Yubikey to test with (I have a differerent key that uses the FIDO/U2F standard), but if you change this section of sign() in Tester.pm:


    # generate the signature
    my $sign_data = pack 'a32 a N a32',
        sha256($app_id),            # 32 byte SHA256 application parameter
        chr(0x00),                  # 1 byte user presence
        $counter,                   # 4 byte counter
        sha256($client_data);       # 32 byte SHA256 of client data JSON

If you change 0x00 to 0x01 for the user presence field does it work?

**Comment by mschout** [comment]
I installed libu2f-server and tested against your v0.43 with above changes and it works now.

Fixed in v0.03, just uploaded to CPAN.

**Comment by guimard** [comment]
Thanks a lot !
This 1-bit is called "User presence" in FIDO spec and required. So Authen::U2F should also reject `0x00` signature.
